### PR TITLE
Remove track/user from metrics uniqueness key

### PR DESCRIPTION
**Pull request by ErikSchierboom**

### CODE REVIEW

**Comment by iHiD** [comment]
I'll leave this to be merged along with you fixing the others simultaneously to stop us creating a lot of duplicates.

**Comment by ErikSchierboom** [comment]
These commands fixup some of the issues regarding github metrics:

```ruby
Metrics::MergePullRequestMetric.group(:params).having('count(*) > 1').count do |m|
  Metrics::MergePullRequestMetric.where(params: m.params, user_id: nil).delete_all
end

Metrics::OpenPullRequestMetric.group(:params).having('count(*) > 1').count do |m|
  Metrics::OpenPullRequestMetric.where(params: m.params, user_id: nil).delete_all
end

Metrics::OpenIssueMetric.group(:params).having('count(*) > 1').count do |m|
  Metrics::OpenIssueMetric.where(params: m.params, user_id: nil).delete_all
end
```

After running this, the following queries all return `true`:

```ruby
Metrics::OpenPullRequestMetric.count == Github::PullRequest.count 
=> true

Metrics::MergePullRequestMetric.count == Github::PullRequest.merged.count
=> true

Metrics::OpenIssueMetric.count == Github::Issue.count
=> true

**Comment by ErikSchierboom** [comment]
The mentoring metrics contain an issue for private mentoring request, which will be fixed in this PR. Once the PR has been merged, we can run:

```ruby
ActiveRecord::Base.transaction(isolation: Exercism::READ_COMMITTED) do
  Mentor::Request.includes(:track, :student).where(external: true).find_each {|r| Metric::Create.(:request_mentoring, r.created_at, track: r.track, user: r.student, request: r) }
end
```

Finished mentoring count is ok.

```ruby
Metrics::FinishMentoringMetric.count == Mentor::Discussion.finished_for_student.count
=> true
```

**Comment by ErikSchierboom** [comment]
We can correct the existing records as follows:

```ruby
Metric.find_each do |m|
  m.update(uniqueness_key: m.uniqueness_key.gsub(/\|\d+\|\d+\|/, '|'))
end
```

An alternative would be to run a big SQL update query.

**Comment by ErikSchierboom** [comment]
The alternative SQL query is:

```sql
UPDATE metrics SET uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1));
```

This even works on records that have already been modified.

File: app/commands/metric/queue.rb [hunk 1]
```diff
@@ -4,9 +4,15 @@ class Metric::Queue
   initialize_with :type, :occurred_at, :attributes
 
   def call
+    return if user&.ghost?
+    return if user&.system?
```
Comment (ErikSchierboom) [inline app/commands/metric/queue.rb:8]: Would you merge these two into one line? I felt this was marginally more clearer, but 🤷

File: app/commands/mentor/request/accept_external.rb [hunk 2]
```diff
@@ -21,6 +22,14 @@ def call
           )
         end
       end
+
+      private
+      def log_metric!(request)
+        Metric::Queue.(:request_mentoring, request.created_at, request:, track:, user:)
```
Comment (iHiD) [inline app/commands/mentor/request/accept_external.rb]: I think this should be a separate metric "RequestPrivateMentoring"

Comment (ErikSchierboom) [reply app/commands/mentor/request/accept_external.rb]: https://github.com/exercism/website/pull/2768/commits/6a160264967a3ce75494262a07f300c4c9553c9a

I'll need to run some queries to fixup the existing metrics.

**Comment by ErikSchierboom** [comment]
```ruby
Metrics::OpenIssueMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")
Metrics::OpenPullRequestMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")
Metrics::MergePullRequestMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")

Metrics::RequestMentoringMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")
Metrics::FinishMentoringMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")

Metrics::PublishSolutionMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")
Metrics::CompleteSolutionMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")
Metrics::SubmitSolutionMetric.update_all("uniqueness_key = CONCAT(SUBSTRING_INDEX(uniqueness_key, '|', 1), '|', SUBSTRING_INDEX(uniqueness_key, '|', -1))")
```

**Comment by ErikSchierboom** [comment]
```ruby
Metric.where(user_id: 720_036).delete_all
Metric.where(user_id: 1).delete_all
```

File: app/commands/metric/queue.rb (hunk 1 above)
Comment (iHiD) [reply app/commands/metric/queue.rb:8]: This is good.

**Review by iHiD — APPROVED** [review]

**Pull request merged**